[JENKINS-67396] Update link and breadcrumb dropdowns

janfaracik opened the pull request:
https://user-images.githubusercontent.com/43062514/146660255-662a6793-3385-439c-8414-bba274548f51.mov

See [JENKINS-67396](https://issues.jenkins.io/browse/JENKINS-67396)

### What's changed

* The way link and breadcrumb dropdowns are rendered has changed, the chevron now belongs to the link/breadcrumb which means it can propagate hover events to its parent (essentially it'll behave and look better)
* The appearance of the chevron has been updated (now using Ionicons) and it now has some subtle animations
* Added dropdowns back for job build numbers - thanks Diego and Tim (https://issues.jenkins.io/browse/JENKINS-67396) 

### Proposed changelog entries

* Update link and breadcrumb dropdowns.

<!-- Comment: 
The changelogs will be integrated by the core maintainers after the merge.  See the changelog examples here: https://www.jenkins.io/changelog/ -->

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [ ] (If applicable) Jira issue is well described
- [ ] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@mention

<!-- Comment:
If you need an accelerated review process by the community (e.g., for critical bugs), mention @jenkinsci/code-reviewers
-->

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

janfaracik committed:
WIP replacement of link dropdowns

janfaracik committed:
Working build - chevron menus now dead, shiny new ones working <3

janfaracik committed:
Final cleanup hopefully :sparkles:

janfaracik committed:
Update tabs.less

janfaracik committed:
Fix missing chevron on user name in header

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik committed:
Update headerContent.jelly

janfaracik committed:
Cleanup header container, use native sticky when possible

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik committed:
Reset stylings

daniel-beck commented:
Is this compatible with "no op" separators as e.g. used by https://github.com/jenkinsci/jenkins/pull/6068 (I'm sure I've seen them elsewhere before too!)?

daniel-beck commented:
The clickable area seems kinda small.

* The "separator link" (for `ModelObjectWithChildren`) is no longer clickable over the full height of the breadcrumb bar, it was before.
* The "model link" is probably around the same size as before, but the new styling emphasizes this further. It probably makes sense for the entire right area of the highlighted area to be clickable?

Good news: This works with agents on `/computer`, users on `/securityRealm/` (if Jenkins user database).

Something I noticed when trying this PR build: There is no longer a tooltip on the "Build" list view column's button. Probably unrelated, but don't have time today to check actual builds.

NotMyFault commented:
I noticed, that headers are styled differently across components. For example, `/manage` or `/configure` displays
![](https://i.imgur.com/YTyi6Oo.png)

But `/pluginManager` displays
![](https://i.imgur.com/wpTmYuZ.png)
The difference is marginal, but definitely smaller.

janfaracik committed:
Fix breadcrumbs not stickying (position: sticky is a pain), fix header height changing (chrome is a pain)

janfaracik committed:
Fixed the footer being misplaced (footers are a pain)

janfaracik committed:
Increase hit target size of chevrons

janfaracik committed:
Fix missing tooltip for build column

janfaracik commented:
> Is this compatible with "no op" separators as e.g. used by #6068 (I'm sure I've seen them elsewhere before too!)?

Not too sure what you mean by "no op" separators, sorry!

> The clickable area seems kinda small.
> 
> * The "separator link" (for `ModelObjectWithChildren`) is no longer clickable over the full height of the breadcrumb bar, it was before.
> * The "model link" is probably around the same size as before, but the new styling emphasizes this further. It probably makes sense for the entire right area of the highlighted area to be clickable?
> 
> Good news: This works with agents on `/computer`, users on `/securityRealm/` (if Jenkins user database).
> 
> Something I noticed when trying this PR build: There is no longer a tooltip on the "Build" list view column's button. Probably unrelated, but don't have time today to check actual builds.

Good suggestions :) I've increased the hitbox for the arrow so its quite a bit larger now, essentially fills the right side of the link now so should make it a fair bit easier to click.

> I noticed, that headers are styled differently across components. For example, `/manage` or `/configure` displays ![](https://camo.githubusercontent.com/4d8e907aed3eaac72bc4047adff9fd54a1c7669748c6a3248032ce7062948134/68747470733a2f2f692e696d6775722e636f6d2f59547969364f6f2e706e67)
> 
> But `/pluginManager` displays ![](https://camo.githubusercontent.com/fa1826dd9d493b845ba42d746e4947b580e81f553a0426316b592fb8f4235da2/68747470733a2f2f692e696d6775722e636f6d2f7770546d59755a2e706e67) The difference is marginal, but definitely smaller.

Great spot! Think that's a bug with Chrome, fixed it by setting the height on the child element too.

timja commented:
Before we ship this we’ll need a green bom and ATH build

timja committed:
Merge branch 'master' into new-breadcrumbs

timja commented:
> Is this compatible with "no op" separators as e.g. used by #6068 (I'm sure I've seen them elsewhere before too!)?
I combined the two branches and everything seems to work perfect.

https://github.com/jenkinsci/jenkins/compare/master...timja:combine-breadcrumbs?expand=1

![image](https://user-images.githubusercontent.com/21194782/147371104-f6069e24-951e-4575-bca2-6a05469c8f15.png)

![image](https://user-images.githubusercontent.com/21194782/147371110-72dc4bcc-2e7e-407a-a1c7-4467c78be9fe.png)

Checked dark theme too 👍

NotMyFault reviewed (changes requested):
The headers are looking good now. I tested a few other components using dropdowns and noticed that the project's build history drop down is behind lower table entries:
![](https://i.imgur.com/24I2XpC.png)
Other dropdowns like for your personal build history table, permalinks etc are looking fine.

timja reviewed (dismissed):
ATH failures unfortunately https://github.com/jenkinsci/acceptance-test-harness/pull/729

janfaracik committed:
Update breadcrumbs.less

janfaracik commented:
> The headers are looking good now. I tested a few other components using dropdowns and noticed that the project's build history drop down is behind lower table entries:
> ![](https://camo.githubusercontent.com/63519cd733537ac1b23e34b337a1188b14838a846d8086a8b9743edc4625c0aa/68747470733a2f2f692e696d6775722e636f6d2f323449325870432e706e67)
> Other dropdowns like for your personal build history table, permalinks etc are looking fine.

Great find, should be fixed now :) 

<img width="353" alt="image" src="https://user-images.githubusercontent.com/43062514/147372390-fc1040aa-73c3-4031-b61d-f3132e4eace4.png">

---

I'll take a look at the tests now

uhafner reviewed (changes requested):
The new UI looks good! 

Two minor things: 
- in the table, the size of an element is now computed without the chevron image. If the text fills the whole column then the chevron pops up in the next column now. 
  <img width="385" alt="Bildschirmfoto 2021-12-27 um 16 14 47" src="https://user-images.githubusercontent.com/503338/147484635-3ec91ba1-48df-4a3e-8b58-0d89fc4655e0.png">
-  The chevron icon appears for empty elements as well. In the old style no chevron was shown if there are no elements to show:
<img width="1352" alt="Bildschirmfoto 2021-12-27 um 16 14 22" src="https://user-images.githubusercontent.com/503338/147484711-4ff97212-f0da-489f-b00a-4058a46e4ba0.png">

While I like the new design it makes the currently broken user experience more prominent now: We have 2 different chevrons right after each other (one with an arrow down, one with an arrow right). This will confuse users a lot. It would be much nicer if we could combine these two menus into a single one.

janfaracik commented:
> The new UI looks good!
> 
> Two minor things:
> 
> * in the table, the size of an element is now computed without the chevron image. If the text fills the whole column then the chevron pops up in the next column now.
>   <img alt="Bildschirmfoto 2021-12-27 um 16 14 47" width="385" src="https://user-images.githubusercontent.com/503338/147484635-3ec91ba1-48df-4a3e-8b58-0d89fc4655e0.png">
> * The chevron icon appears for empty elements as well. In the old style no chevron was shown if there are no elements to show:
> 
> <img alt="Bildschirmfoto 2021-12-27 um 16 14 22" width="1352" src="https://user-images.githubusercontent.com/503338/147484711-4ff97212-f0da-489f-b00a-4058a46e4ba0.png">
> 
> While I like the new design it makes the currently broken user experience more prominent now: We have 2 different chevrons right after each other (one with an arrow down, one with an arrow right). This will confuse users a lot. It would be much nicer if we could combine these two menus into a single one.

I've marked the PR as draft while I look at these comments + tests - I'll try and have something up soon to address them - thanks everyone!

janfaracik committed:
Working build - first attempt at combining breadcrumb dropdowns

janfaracik committed:
Minor styling changes to breadcrumb bar

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik committed:
Javarey menus beat loading them twice

timja commented:
How are we looking now @janfaracik ?

janfaracik committed:
Fix config page breadcrumb dropdown, adjust js doc

janfaracik committed:
Update Javadoc

janfaracik commented:
> How are we looking now @janfaracik ?

Better! I've undrafted the PR now :) 

I've updated the breadcrumb bar following @uhafner's comments - 

<img width="335" alt="image" src="https://user-images.githubusercontent.com/43062514/147886785-89ad2d11-b418-4c25-bfa8-81b92837f57b.png">

Breadcrumb items no longer have two chevrons, instead they have the one visible one which now contains the context menu items as well as the children menu items.

This also extends to normal links too - 

<img width="227" alt="image" src="https://user-images.githubusercontent.com/43062514/147886808-186d74d4-548c-4b98-847c-4e7e41e0dc30.png">

(Pardon the separator - it's been replaced in the upcoming YUI menu replacement PR)

---

The chevron popping up in the next column under very small widths is a tricky one, would need to spend more time looking into as.

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

timja commented:
The system 'configuration' breadcrumb link, links to `rootUrl/0.0`? On ci.jenkins.io the link doesn't lead anywhere.

![image](https://user-images.githubusercontent.com/21194782/147916291-bdbccd70-995c-4af3-91d1-eddef0d9a1ed.png)

Apart from that it looks a lot better to me, wouldn't be surprised if it hits some edge cases but in the general case this is far nicer.

janfaracik committed:
Fix breadcrumb link

janfaracik commented:
> The system 'configuration' breadcrumb link, links to `rootUrl/0.0`? On ci.jenkins.io the link doesn't lead anywhere.
> 
> ![image](https://user-images.githubusercontent.com/21194782/147916291-bdbccd70-995c-4af3-91d1-eddef0d9a1ed.png)
> 
> Apart from that it looks a lot better to me, wouldn't be surprised if it hits some edge cases but in the general case this is far nicer.

My bad! Had a bracket in the wrong spot - it should now link to the Configuration page :)

timja commented on core/src/main/resources/lib/form/breadcrumb-config-outline.jelly:
```diff
@@ -29,5 +29,5 @@ THE SOFTWARE.
 
   <st:adjunct includes="lib.form.breadcrumb-config-outline.init"/>
-  <l:breadcrumb title="${%configuration}" href="${rootURL/configure}" id="inpage-nav" />
+  <l:breadcrumb title="${%configuration}" href="${rootURL}/configure" id="inpage-nav" />
```
should probably be a capital I think? looks weird without
```suggestion
  <l:breadcrumb title="${%Configuration}" href="${rootURL}/configure" id="inpage-nav" />
```

janfaracik committed:
Update core/src/main/resources/lib/form/breadcrumb-config-outline.jelly

Thanks @timja

Co-authored-by: Tim Jacomb <21194782+timja@users.noreply.github.com>

timja commented:
From my POV the ATH tests are next, but requested review from others again

daniel-beck commented:
> Not too sure what you mean by "no op" separators, sorry!

Existing separators have an associated menu/action (the children menu), and AFAICT that PR would introduce separators without a menu.

daniel-beck commented on core/src/main/java/jenkins/model/ModelObjectWithPageMenu.java:
```diff
@@ -0,0 +1,44 @@
+ * @see ModelObjectWithContextMenu
+ * @since TODO - Provide version
+ */
+public interface ModelObjectWithPageMenu extends ModelObject {
```
This looks like a UI implementation detail better not reflected in the type hierarchy, why is this entire thing not just in `Functions`?

daniel-beck commented:
Some feedback from running the latest build:

* There is a "dropdown" animation even on no-op dividers, like the one after "Plugin Manager" at `/pluginManager/`. This wasn't great before, but the new animation makes it worse. Note that there are actual situations in which a menu should appear but doesn't (e.g. session timeout), so this behavior is usually indicative of a problem.
* While merging the two menus makes the children menu more easily discoverable, I'm not confident it's a great long-term solution, especially given the large and variable number of entries in either part of the menu. (Ideally it would therefore not be reflected in the type hierarchy per my previous comment.)
* The divider line looks terrible. Could we at least do emdashes or something, that appears as a line? Or a magic value that gets rendered as an `<hr/>`?
* Ideally we'd add proper support for non-link entries (i.e. no highlighting either) in the menus, that could be applied to the titles in the grouped "Manage Jenkins" menu as well
* I don't know whether this is new here, but the Jenkins 2.0 config form for jobs has a badly aligned buttons block: <img width="457" alt="Screenshot 2022-01-03 at 14 48 10" src="https://user-images.githubusercontent.com/1831569/147937992-684c662f-5e92-410d-bd53-ccf64669f10d.png">
* It's odd to see log output moving around when hovering links from `ConsoleNote`s, like `a1` in "Building remotely on `a1` in workspace /tmp/a1/workspace/fff".
* Such build log output is no longer friendly to copy & paste, because ConsoleNote model links as in the previous item are now followed by two newlines (Firefox 95.0.2 on Mac OS), so it's actually

      Building remotely on a1
      
       in workspace /tmp/a1/workspace/fff

uhafner reviewed (changes requested):
The combination of the 2 chevron icons looks good now. Also the breadcrumb items. 

Some functionality is still broken: the breadcrumbs menu for builds is now empty. And for empty breadcrumb menus still a chevron is shown.

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik committed:
Update separator design

janfaracik committed:
Make menu headers/separators non-clickable

janfaracik committed:
Fix build breadcrumb menu not appearing

janfaracik committed:
Update styling slightly

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik committed:
Improve model links for console/text no longer moves + text now copies correctly on firefox

janfaracik committed:
Breadcrumb chevrons without menus no longer have hover/active states, hide non menu separators if theyre last

janfaracik commented:
> Some feedback from running the latest build:
> 
> * There is a "dropdown" animation even on no-op dividers, like the one after "Plugin Manager" at `/pluginManager/`. This wasn't great before, but the new animation makes it worse. Note that there are actual situations in which a menu should appear but doesn't (e.g. session timeout), so this behavior is usually indicative of a problem.

I've updated this so that separators (chevrons) without menus no longer have hover/active states and if the separator without a menu is the last element it'll be hidden. 

> * While merging the two menus makes the children menu more easily discoverable, I'm not confident it's a great long-term solution, especially given the large and variable number of entries in either part of the menu. (Ideally it would therefore not be reflected in the type hierarchy per my previous comment.)

I don't really have an opinion on this part - I do think the menus could do with better organisation however and that overly long menus aren't too helpful/productive.

> * The divider line looks terrible. Could we at least do emdashes or something, that appears as a line? Or a magic value that gets rendered as an `<hr/>`?
> * Ideally we'd add proper support for non-link entries (i.e. no highlighting either) in the menus, that could be applied to the titles in the grouped "Manage Jenkins" menu as well

Ah I didn't spend too much time on the divider as I'm planning on replacing it fairly soon once I've got the YUI menu/tooltip branch polished. 

I've updated this PR to have a new separator + I've removed hover/active events for separators/headers.

<img width="243" alt="image" src="https://user-images.githubusercontent.com/43062514/149213198-d53536ca-d2e3-4c24-acdc-b7ef837a3a62.png">

> * I don't know whether this is new here, but the Jenkins 2.0 config form for jobs has a badly aligned buttons block:

That'll be replaced in the upcoming forms updates :)

> * It's odd to see log output moving around when hovering links from `ConsoleNote`s, like `a1` in "Building remotely on `a1` in workspace /tmp/a1/workspace/fff".
> * Such build log output is no longer friendly to copy & paste, because ConsoleNote model links as in the previous item are now followed by two newlines (Firefox 95.0.2 on Mac OS), so it's actually
>   ```
>   Building remotely on a1
>   
>    in workspace /tmp/a1/workspace/fff
>   ```

Great spots! I've updated this so that the text no longer moves + put in a small workaround for Firefox for copying console text.

https://user-images.githubusercontent.com/43062514/149213398-f0409b79-536c-495f-a586-3584a25818dc.mov

> The combination of the 2 chevron icons looks good now. Also the breadcrumb items.
> 
> Some functionality is still broken: the breadcrumbs menu for builds is now empty. And for empty breadcrumb menus still a chevron is shown.

I've fixed these, if there's any other breadcrumbs that have menus that shouldn't let me know :) 

---

Thanks everybody for the reviews - really appreciated.

daniel-beck commented:
> Truthfully I don't know a whole bunch about how Jenkins' endpoints are built - I wanted a new endpoint for each object that has a context menu/children menu to consolidate them into one call so went about it this way, is there a better way of going about it/any examples of something similar?

Ah, OK. My perception was that this menu is "just" a combined children/context menu, rather than a new thing superseding the existing ones. The Javadoc also reads like that. So adding a new type would be unnecessary, as its current functionality could as easily be in a helper method in `Functions`. If the idea is that the children/context distinction is being retired, that makes more sense.

> I don't really have an opinion on this part - I do think the menus could do with better organisation however and that overly long menus aren't too helpful/productive.

Right, hence my point that we used to have two menus, each with a small to large number of items, and are combining them into one with a medium to huge number of items instead. Nesting basically doesn't exist (I think Manage Jenkins on the Jenkins root menu is the only submenu, and that one is special anyway), and there's no visibility controls hiding less relevant/used items (and those would be difficult to coordinate too).

> I've updated this PR to have a new separator + I've removed hover/active events for separators/headers.

Screenshot looks great, I look forward to running this!

> Great spots! I've updated this so that the text no longer moves + put in a small workaround for Firefox for copying console text.

Thanks! Is this something that's feasible to autotest somehow, any idea? I expect this to be of some importance to users and it feels like a very lucky discovery we could have easily missed.

NotMyFault reviewed (approved):
My and other outstanding concerns have been addressed so far, I was looking over it again earlier, but didn't find something new.

Could we fire up an ATH test? Afaik that one is still on the todo list.

timja commented:
> Could we fire up an ATH test? Afaik that one is still on the todo list.

It's here: https://github.com/jenkinsci/acceptance-test-harness/pull/730

and failing atm

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

timja committed:
Merge branch 'master' into new-breadcrumbs

timja commented:
@daniel-beck do you want to review this again?

The ATH tests should be fixed today, hopefully the last run is going through atm.

daniel-beck commented:
Unless and until there's a useful implementation of the new class that demonstrates its benefits, I'd strongly prefer the menus of `WithChildren` and `WithContextMenu` be merged in a utility function, and no new type be introduced that may just become a maintenance burden later on.

timja commented:
> Besides that, it looks like my previous comment remains unanswered so far?

specifically?

daniel-beck commented:
> specifically?

I'm touching on the following topics:

* New type, which seems unnecessary; unless it's conceptually different from just a combined menu (amended by my new comment that a new type should have a reference implementation demonstrating the benefits)
* Potentially excessive length of combined menus when previously, two separate menus were fine
* Autotesting copy & paste behavior of console notes that regressed in this PR at first

uhafner reviewed (approved):
From the UI perspective it look fine now!

janfaracik commented:
> Unless and until there's a useful implementation of the new class that demonstrates its benefits, I'd strongly prefer the menus of `WithChildren` and `WithContextMenu` be merged in a utility function, and no new type be introduced that may just become a maintenance burden later on. 
 
> * New type, which seems unnecessary; unless it's conceptually different from just a combined menu (amended by my new comment that a new type should have a reference implementation demonstrating the benefits)

I think the issue may be down to my inexperience with the Jenkins’ codebase - essentially I want a new endpoint for each object that has a context menu/children menu to consolidate them into one call, as calling two endpoints to load the menu was really slow/unresponsive. 

If there’s a better way of adding that endpoint than what I’m currently doing I’m definitely up for doing it @daniel-beck @timja

> * Potentially excessive length of combined menus when previously, two separate menus were fine

I am in agreement with @uhafner on reducing the amount of chevrons in the breadcrumb bar - the two chevrons before were a bit confusing/inconsistent with the rest of Jenkins. 

I’m of the view that the dropdown menus are something that should definitely be looked at again in the future, I’ve got a branch in progress at the moment that’ll revamp their appearance/tech (demoed at the UX sig) and I’d definitely be up for a wider discussion on how to improve the menus contents.

> * Autotesting copy & paste behavior of console notes that regressed in this PR at first

I’ll add that if I can get it working on the ac-test-harness :thumbsup:

daniel-beck commented:
@janfaracik Thanks for your response!

> I think the issue may be down to my inexperience with the Jenkins’ codebase - essentially I want a new endpoint for each object that has a context menu/children menu to consolidate them into one call, as calling two endpoints to load the menu was really slow/unresponsive.

Good point, that should be unnecessary. I'll try to think of something. Unsure how `@Restricted` on super types would impact plugin development, might be a possible solution.

timja reviewed (commented):
FYI bom and ATH are passing now on this

timja commented:
conflicts

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik commented:
> I'm not sure if that is possible: can we remove the `ModelObjectWithPageMenu` from this PR and integrate only the UI changes? Then we could move this discussion to a separate thread.
> 
> In this new thread we can discuss more conceptually how we would like to use those menus in the future. I think when talking about these menus we should also discuss the sidebar items as well. I think both elements have been introduced in Jenkins by simply adding them on the Java code side without really thinking about how this will look like when many plugins use them. I would rather prefer to have the control over these element in core. So we don't get an explosion of side bar or menu elements just because it is possible. As a rule of thumb in UI design: no context menu should have more than 7 items 😉

I'll get working on that 👍 I'll update this PR when that's done.

janfaracik committed:
Restore separate breadcrumb dropdowns

janfaracik committed:
Remove IE compat

janfaracik committed:
Remove log

janfaracik committed:
Update core/src/main/java/jenkins/model/ModelObjectWithContextMenu.java

Co-authored-by: Tim Jacomb <21194782+timja@users.noreply.github.com>

timja reviewed (approved):
LGTM, tested via core pr tester

uhafner reviewed (approved):
Thanks! This is now a good starting point for improvements of the different menu types.

NotMyFault reviewed (approved):
Works great, nice to see that IE compatibility is no longer a blocker :P

janfaracik commented:
Any further thoughts on this @uhafner @daniel-beck?

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

uhafner commented:
Everything looks fine! Ready to go 🚀

daniel-beck commented:
Is there a reason for the animations? Since the (fairly small) clickable areas move around, that makes it unnecessarily difficult to correct dexterity mistakes, like clicking the children menu instead of the dropdown menu.

Old:
https://user-images.githubusercontent.com/1831569/155569630-28f0bc0a-d362-4785-8bdb-3954514ac6a0.mov

New:
https://user-images.githubusercontent.com/1831569/155569638-ed5aa47f-b148-46d5-91dc-d18ff3387bfe.mov


Additionally, it's no longer possible to directly open a menu without hovering the element first, so all mouse movements are L shaped. (How does this work on touchscreens? Not saying we were great before, but are we getting worse?)


Old:
https://user-images.githubusercontent.com/1831569/155569175-b979ba0e-35f9-4539-ae7d-78ddc10e39c0.mov

New:
https://user-images.githubusercontent.com/1831569/155569190-0d997c1f-4924-48ab-bf7d-45dd00c18e5d.mov

daniel-beck commented:
I also noticed this (on hover):
<img width="765" alt="Screenshot 2022-02-24 at 17 33 18" src="https://user-images.githubusercontent.com/1831569/155573936-e83c3a21-19fd-4c04-9cc2-5751b0f36c24.png">
I guess that's the price for it being copy & pastable cleanly?

daniel-beck commented:
The new appearance and lack of hover effect on the section headers in the "Manage Jenkins" dropdown is great! Thanks for taking care of that.

timja commented:
@janfaracik you able to look at Daniel's comments please?

janfaracik committed:
Update so chevrons appear for touch screen users, increase background contrast

janfaracik commented:
> Is there a reason for the animations? Since the (fairly small) clickable areas move around, that makes it unnecessarily difficult to correct dexterity mistakes, like clicking the children menu instead of the dropdown menu.
> 
> Old: https://user-images.githubusercontent.com/1831569/155569630-28f0bc0a-d362-4785-8bdb-3954514ac6a0.mov
> 
> New: https://user-images.githubusercontent.com/1831569/155569638-ed5aa47f-b148-46d5-91dc-d18ff3387bfe.mov

For the breadcrumbs it stems from not being a fan of the spacing and hotboxes, the expanding animation allows for slightly larger hitboxes. Ideally though we'll be able to reduce the chevrons in the breadcrumb bar when we have the menu discussion.

For the other links though it's just a stylistic decision so that there aren't large gaps between links for the chevrons, e.g. on the build history page.

<img width="238" alt="image" src="https://user-images.githubusercontent.com/43062514/156055188-111f5d92-6c85-43e7-b92a-53079d732865.png">

> Additionally, it's no longer possible to directly open a menu without hovering the element first, so all mouse movements are L shaped. (How does this work on touchscreens? Not saying we were great before, but are we getting worse?)
> 
> Old: https://user-images.githubusercontent.com/1831569/155569175-b979ba0e-35f9-4539-ae7d-78ddc10e39c0.mov
> 
> New: https://user-images.githubusercontent.com/1831569/155569190-0d997c1f-4924-48ab-bf7d-45dd00c18e5d.mov

That's a really good point, did a bit of Googling around this and it turns out there's now CSS queries for touchscreens. I've updated it so that chevrons are always visible when a user is using a touchscreen. It looks like so:

<img width="1228" alt="image" src="https://user-images.githubusercontent.com/43062514/156055283-8b632fb0-d5f9-4535-a080-ed387bdd2c0c.png">

> I also noticed this (on hover):
> <img alt="Screenshot 2022-02-24 at 17 33 18" width="765" src="https://user-images.githubusercontent.com/1831569/155573936-e83c3a21-19fd-4c04-9cc2-5751b0f36c24.png">
> I guess that's the price for it being copy & pastable cleanly?

Ah the background for the link is supposed to be darker, but with the forms update the console now has a darker background so they're blending in.

Updated it so it looks like so:

<img width="322" alt="image" src="https://user-images.githubusercontent.com/43062514/156055788-5059c336-8113-49fc-82d5-bd4c5fac38d9.png">

daniel-beck commented:
> Ideally though we'll be able to reduce the chevrons in the breadcrumb bar when we have the menu discussion.

I see. I think the current status is OK as a placeholder.

The second item of feedback still applies in general though, right? There'll be lots of L-shaped cursor movements since the menu button is to the right of the hover area that makes it appear?

> Ah the background for the link is supposed to be darker, but with the forms update the console now has a darker background so they're blending in.
> 
> Updated it so it looks like so:

Thanks, that makes it look deliberate rather than broken. Definitely a lot better than text moving around or not having a clean copy & paste!

janfaracik committed:
Merge branch 'master' into new-breadcrumbs

janfaracik committed:
Increase hitbox

janfaracik commented:
> The second item of feedback still applies in general though, right? There'll be lots of L-shaped cursor movements since the menu button is to the right of the hover area that makes it appear?

I've just increased the hitbox on the right by a fair amount so that users can quickly scan their cursor down the links to show the menus, behaves very similar to the old way now, hope that makes it better.

https://user-images.githubusercontent.com/43062514/156077558-aab4f85e-9e44-489a-9614-b97da1eb9d88.mov

daniel-beck commented:
@janfaracik Looks great! I'll give it a try in a day or two , really looking forward to this.

timja commented:
Need to run this back through ATH just to be safe

NotMyFault commented:
The updated chevrons are looking much better now 👍🏻
 
Only new found I have is that the drop down to manage an agent almost vanishes, if the agent is offline:
![](https://i.imgur.com/HY2JirL.png)

This is on 2.337 for a better comparison:
![](https://i.imgur.com/6ivUrq0.png)

timja commented:
ATH passed https://github.com/jenkinsci/acceptance-test-harness/pull/749

janfaracik commented:
> The updated chevrons are looking much better now 👍🏻
> 
> Only new found I have is that the drop down to manage an agent almost vanishes, if the agent is offline: ![](https://camo.githubusercontent.com/84a209ad760fc78be9b06d5d9743fcba2c7b7e0a84ca8c77330da73e062d41c4/68747470733a2f2f692e696d6775722e636f6d2f4859324a69724c2e706e67)
> 
> This is on 2.337 for a better comparison: ![](https://camo.githubusercontent.com/113592cea91df7723ff3aa1fed323bd03bd090bf435506b252f183ebe7d1b45b/68747470733a2f2f692e696d6775722e636f6d2f366976557271302e706e67)

Hmm a few potential fixes for that - 

1. We could add some space between the two bits of text e.g.
<img width="172" alt="image" src="https://user-images.githubusercontent.com/43062514/156369286-2ebd3d40-654c-429f-a07d-3d9dd7d0488a.png">

2. Move the status to the right, e.g.
<img width="341" alt="image" src="https://user-images.githubusercontent.com/43062514/156368999-32d37760-855e-4038-a31d-cbc331286490.png">

3. Move the status below

Any thoughts?

NotMyFault commented:
> > The updated chevrons are looking much better now 👍🏻
> > Only new found I have is that the drop down to manage an agent almost vanishes, if the agent is offline: ![](https://camo.githubusercontent.com/84a209ad760fc78be9b06d5d9743fcba2c7b7e0a84ca8c77330da73e062d41c4/68747470733a2f2f692e696d6775722e636f6d2f4859324a69724c2e706e67)
> > This is on 2.337 for a better comparison: ![](https://camo.githubusercontent.com/113592cea91df7723ff3aa1fed323bd03bd090bf435506b252f183ebe7d1b45b/68747470733a2f2f692e696d6775722e636f6d2f366976557271302e706e67)
> 
> Hmm a few potential fixes for that -
> 
> 1. We could add some space between the two bits of text e.g.
> 
> <img alt="image" width="172" src="https://user-images.githubusercontent.com/43062514/156369286-2ebd3d40-654c-429f-a07d-3d9dd7d0488a.png">
> 
> 2. Move the status to the right, e.g.
> 
> <img alt="image" width="341" src="https://user-images.githubusercontent.com/43062514/156368999-32d37760-855e-4038-a31d-cbc331286490.png">
> 
> 3. Move the status below
> 
> Any thoughts?

Aligning it on the right side sounds good. That way, offline agents with a longer name don't look out of place too:
![](https://i.imgur.com/fkbFqfv.png)

janfaracik committed:
Update node status to align to the right

NotMyFault reviewed (approved):
Reapproving; (my) concerns have been addressed 🎉

timja commented:
@daniel-beck will you have time to check this soon?

daniel-beck commented:
@timja  Should be doable this week. Thanks for checking!

daniel-beck commented on core/src/main/resources/jenkins/views/JenkinsHeader/headerContent.jelly:
```diff
@@ -1,6 +1,6 @@
 <?jelly escape-by-default='true'?>
 <j:jelly xmlns:j="jelly:core" xmlns:st="jelly:stapler" xmlns:d="jelly:define" xmlns:l="/lib/layout" xmlns:i="jelly:fmt" xmlns:x="jelly:xml">
-    <header id="header" class="page-header">
+    <header id="page-header" class="page-header">
```
This looks like it could easily break some stuff inadvertently for no real reason?

daniel-beck commented on core/src/main/resources/lib/form/breadcrumb-config-outline.jelly:
```diff
@@ -32,5 +32,5 @@ THE SOFTWARE.
 
   <st:adjunct includes="lib.form.breadcrumb-config-outline.init"/>
-  <l:breadcrumb title="${attrs.title?:'%configuration'}" id="inpage-nav" />
+  <l:breadcrumb title="${attrs.title?:'%configuration'}" href="${rootURL}/configure" id="inpage-nav" />
```
This makes no sense given this is a general purpose element.

daniel-beck commented on core/src/main/resources/lib/form/breadcrumb-config-outline.jelly:
```diff
@@ -32,5 +32,5 @@ THE SOFTWARE.
 
   <st:adjunct includes="lib.form.breadcrumb-config-outline.init"/>
-  <l:breadcrumb title="${attrs.title?:'%configuration'}" id="inpage-nav" />
+  <l:breadcrumb title="${attrs.title?:'%configuration'}" href="${rootURL}/configure" id="inpage-nav" />
```
This makes no sense given this is a general purpose element.

janfaracik replied:
I can't quite remember to be honest! I've reverted this 👍

daniel-beck commented on core/src/main/resources/lib/layout/breadcrumbs.js:
```diff
@@ -212,15 +134,20 @@ var breadcrumbs = (function() {
-                        if (e.header) {
+                        if (e.type === "HEADER") {
                             e.text = makeMenuHtml(e.icon, e.iconXml, "<span class='header'>" + e.displayName + "</span>");
+                            e.disabled = true;
```
❤️

daniel-beck commented on war/src/main/js/util/page.js:
```diff
@@ -41,7 +41,7 @@ function onWinScroll(callback) {
 
 function pageHeaderHeight() {
-    return elementHeight('#page-head');
+    return elementHeight('#page-header');
```
I am not confident this is correct. Seems used by `config-scrollspy.js` and starting to scroll on a Jenkins 2.0 style job config form is jarring in the PR build.

janfaracik replied:
Good spot! It was missing the addition of the breadcrumb height, fixed now.

daniel-beck reviewed (commented):
Functionality-wise, this seems to be working great now. Thank you for incorporating my feedback.

Some implementation notes.

janfaracik committed:
Remove HREF from breadcrumb

janfaracik committed:
Fix scroll spy header

timja commented:
Anyone want to do a re-check after last few changes?

NotMyFault reviewed (commented):
One of the very recent changes made the console breadcrumbs even worse:

https://user-images.githubusercontent.com/13383509/158257758-6039be22-a8e1-4671-9bc6-23de890f7bd1.mp4

Edit: I merged master/this pr before, that explains the font difference at the console view, but the same applies without as well.

uhafner commented:
I also think that there are still some problems when the links are in text:


https://user-images.githubusercontent.com/503338/158262623-0037b232-05e3-42c5-87ea-8254a39f2f0a.mov

uhafner commented:
Also I must admit that Daniels comment about the L movement still is a usability problem. So in order to get the menu I need to hoover over *Dashboard*, then the icon appears, then I need to carefully navigate right to the appearing icon (some pixels to the left or right open the wrong thing), then I can navigate down. This requires a very careful movement now.

daniel-beck commented:
> This requires a very careful movement now.

Right, it's not great, but per https://github.com/jenkinsci/jenkins/pull/6084#issuecomment-1054670604 both Jan and I seem to agree this is temporary until we figure out a better solution for the dual dropdown menus for breadcrumbs, and I think it's good enough for that.

uhafner commented:
> by extending the hoverable area to the right.

This does not work in the header breadcrumbs there you need to hover over the text and then move to the right to access the menu. So a left-right-down movement.

But I am fine with the changes, I actually do not use the breadcrumb menus very often. Hopefully we find a better way in another iteration that will not require that the texts move horizontally while hovering over an item.

NotMyFault reviewed (approved):
Approving, thanks for your clarification comment, Daniel.

janfaracik committed:
Fix dropdown in console/started by

janfaracik commented:
Great spots guys, cheers! 

Should be fixed now :)

<img width="885" alt="image" src="https://user-images.githubusercontent.com/43062514/158280933-03fa2fbe-2ca9-4f9d-8cae-20a7df04c159.png">

daniel-beck commented:
Has https://github.com/jenkinsci/jenkins/pull/6084#issuecomment-1067302250 been addressed?

timja committed:
Merge branch 'new-breadcrumbs' of github.com:janfaracik/jenkins into new-breadcrumbs

timja committed:
Merge branch 'master' into new-breadcrumbs

timja commented:
> Has [#6084 (comment)](https://github.com/jenkinsci/jenkins/pull/6084#issuecomment-1067302250) been addressed?

It's fixed

Master:
![image](https://user-images.githubusercontent.com/21194782/158337236-7e79f1b4-e719-4d33-b4ae-5525999f41f5.png)

This branch:
![image](https://user-images.githubusercontent.com/21194782/158337273-19d0ca46-fa2f-40c1-8fe8-d257363fc5b0.png)

Hovered:
![image](https://user-images.githubusercontent.com/21194782/158337369-eef5f357-bb92-48ef-9ff8-90fbeaebbbac.png)

timja commented:
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

daniel-beck commented:
In the build history of a job, the hover area of the `#123` overlaps about half the link to the build log to its left on the build status/result icon. This makes it very difficult to directly jump to the build log.

https://user-images.githubusercontent.com/1831569/158356351-14503fac-3534-41b4-be14-df6c7a4484c1.mov

janfaracik committed:
Fix build icon not being clickable

janfaracik commented:
> In the build history of a job, the hover area of the `#123` overlaps about half the link to the build log to its left on the build status/result icon. This makes it very difficult to directly jump to the build log.
> 
>  Untitled.mov

Fixed 👍

daniel-beck commented:
As a plugin developer who has `model-link` links in his plugin, how do I know what additional classes to add to these links for the behavior to be correct? Is that documented anywhere?

janfaracik commented:
> As a plugin developer who has `model-link` links in his plugin, how do I know what additional classes to add to these links for the behavior to be correct? Is that documented anywhere? In the past I've probably copied and pasted `model-link inside` everywhere and that was it. Seeing this PR indicates that's no longer the case?

That's a really good point, I haven't had a chance to try a plugin that uses `model-link` - how are the changes performing on your plugin?

I'll work on some documentation for this to go on dev docs and I'll also add some to the `ui-samples` plugin update.

daniel-beck commented:
> how are the changes performing on your plugin

No specific plugin in mind, that was more of a user story kind of phrasing.

zbynek commented:
@janfaracik Was it intentional to reduce the font weight here? It went from `bold` to `500`, but some fonts render that as normal (tested on Ubuntu which uses Noto Sans font). The table header and the bar above it use font weight 600 which is rendered as bold.
![font-weight](https://user-images.githubusercontent.com/1105305/159867323-549685f6-797d-4df3-b79f-775d705069dc.png)

basil commented:
> Was it intentional to reduce the font weight here? It went from `bold` to `500`

Hi @zbynek, can you please file a ticket in Jira regarding the move from `bold` to `500` if one has not been filed already, noting the steps to reproduce, the elements whose boldness has changed, and the first release which the problem is visible. Thanks!

basil commented:
FTR a Git bisect shows that this causes [JENKINS-68129](https://issues.jenkins.io/browse/JENKINS-68129) / [JENKINS-68097](https://issues.jenkins.io/browse/JENKINS-68097).

basil commented:
> Hi @zbynek, can you please file a ticket in Jira regarding the move from `bold` to `500` if one has not been filed already,

Perhaps [JENKINS-67743](https://issues.jenkins.io/browse/JENKINS-67743) is covering this already?

zbynek commented:
> Perhaps [JENKINS-67743](https://issues.jenkins.io/browse/JENKINS-67743) is covering this already?

That was introduced in a different PR, so I created [JENKINS-68221](https://issues.jenkins.io/browse/JENKINS-68221), but if someone wants to merge the tickets that's fine for me of course.

daniel-beck commented:
Causes [JENKINS-68160](https://issues.jenkins.io/browse/JENKINS-68160).

basil commented:
Hello @janfaracik, there are currently 4 open regressions caused by this change:

- [JENKINS-68097](https://issues.jenkins.io/browse/JENKINS-68097) Credential domain dropdown floats into text (regression in 2.340)
- [JENKINS-68103](https://issues.jenkins.io/browse/JENKINS-68103) Clicking build progress bar no longer jumps to console output (regression in 2.340)
- [JENKINS-68129](https://issues.jenkins.io/browse/JENKINS-68129) Dropdown chevrons no longer separate lists (regression in 2.340)
- [JENKINS-68160](https://issues.jenkins.io/browse/JENKINS-68160) Context menu change causes weird behavior in build history (regression in 2.340)

I fear that these issues will cause disruption to users if delivered in an LTS release. You are in a strong position to be able to resolve these issues efficiently; would you be able to take a look at them?

janfaracik commented:
> * JENKINS-68097

Looking into this now, might require a bit of reworking to the CSS but we'll see.

daniel-beck commented on core/src/main/resources/hudson/views/BuildButtonColumn/column.jelly:
```diff
@@ -29,18 +29,18 @@ THE SOFTWARE.
+                  <j:set var="title" value="${%Schedule_a_task(h.getRelativeDisplayNameFrom(job, itemGroup),it.taskNoun(job))}"/>
+                </j:otherwise>
+              </j:choose>
+              <a id="${id}" tooltip="${title}" class="jenkins-table__button jenkins-table__button--green" href="${href}">
```
Caused [SECURITY-2780](https://www.jenkins.io/security/advisory/2022-06-22/#SECURITY-2781)

daniel-beck commented on core/src/main/java/hudson/console/ModelHyperlinkNote.java:
```diff
@@ -27,7 +27,7 @@ public ModelHyperlinkNote(String url, int length) {
     @Override
     protected String extraAttributes() {
-        return " class='model-link'";
+        return " class='model-link model-link--float'";
```
I thought I had mentioned it before (and may have, elsewhere), but stuff like this is a poor solution, as it will only apply to links generated after this update has been applied. Historical build logs, which are very important for many Jenkins users, will not get this change applied.
